Incorrect suggestion when assigning `Option<T>` to a `T` in a struct pattern

### Code

```Rust
fn main() {
    let a: usize;
    
    Thing {
        thing: a
    } = Thing {
        thing: Some(0)
    };
}

struct Thing {
    thing: Option<usize>
}
```

### Current output

```Shell
error[E0308]: mismatched types
 --> src/main.rs:5:16
  |
2 |     let a: usize;
  |            ----- expected due to this type
...
5 |         thing: a
  |                ^ expected `usize`, found `Option<usize>`
  |
  = note: expected type `usize`
             found enum `Option<usize>`
help: consider using `Option::expect` to unwrap the `Option<usize>` value, panicking if the value is an `Option::None`
  |
5 |         thing: a.expect("REASON")
  |                 +++++++++++++++++

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```

### Desired output

```Shell
error[E0308]: mismatched types
 --> src/main.rs:5:16
  |
2 |     let a: usize;
  |            ----- expected due to this type
...
5 |         thing: a
  |                ^ expected `usize`, found `Option<usize>`
  |
  = note: expected type `usize`
             found enum `Option<usize>`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```

### Rationale and extra context

[Playground Link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=b2398b6eac249beaf0e30a8e24999d2e)

The suggestion to use `.expect("REASON")` on the identifier in the struct pattern is incorrect and doesn't make sense.

### Other cases

```Rust
The same problem is not seen in tuple structs, e.g. the output of the following is fine:

fn main() {
    let a: usize;
    
    Thing(a) = Thing(Some(0));
}

struct Thing(Option<usize>);
```

### Rust Version

```Shell
version 1.95.0 on the playground.
```

### Anything else?

_No response_

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
@@ -2218,6 +2218,19 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expected: Ty<'tcx>,
         found: Ty<'tcx>,
     ) -> bool {
+        // don't suggest missing `.expect()` or `?` in destructuring assignments LHS.
+        // If the immediate parent is an Assign Expr, and the LHS and the RHS of that Expr
+        // overlap with each other, it's guaranteed that the expression came from desugaring
+        // a destructuring assignment.
+        let parent_node = self.tcx.parent_hir_node(expr.hir_id);
+        if let hir::Node::Expr(e) = parent_node
+            && let hir::ExprKind::Assign(lhs, rhs, _) = e.kind
+            && rhs.hir_id == expr.hir_id
+            && lhs.span.overlaps(rhs.span)
+        {
+            return false;
+        }
+
         let ty::Adt(adt, args) = found.kind() else {
             return false;
         };

--- a/tests/ui/destructuring-assignment/dont-suggest-wrong-expect-suggestion.rs
+++ b/tests/ui/destructuring-assignment/dont-suggest-wrong-expect-suggestion.rs
@@ -1,0 +1,19 @@
+//! regression test for <https://github.com/rust-lang/rust/issues/155516>
+
+struct NamedStruct {
+    opt_field: Option<usize>,
+    res_field: Result<usize, ()>,
+}
+
+fn main() {
+    let a: usize;
+    let b: usize;
+
+    NamedStruct {
+        opt_field: a, //~ ERROR mismatched types
+        res_field: b, //~ ERROR mismatched types
+    } = NamedStruct {
+        opt_field: Some(0),
+        res_field: Ok(42_usize),
+    };
+}

--- a/tests/ui/destructuring-assignment/dont-suggest-wrong-expect-suggestion.stderr
+++ b/tests/ui/destructuring-assignment/dont-suggest-wrong-expect-suggestion.stderr
@@ -1,0 +1,27 @@
+error[E0308]: mismatched types
+  --> $DIR/dont-suggest-wrong-expect-suggestion.rs:13:20
+   |
+LL |     let a: usize;
+   |            ----- expected due to this type
+...
+LL |         opt_field: a,
+   |                    ^ expected `usize`, found `Option<usize>`
+   |
+   = note: expected type `usize`
+              found enum `Option<usize>`
+
+error[E0308]: mismatched types
+  --> $DIR/dont-suggest-wrong-expect-suggestion.rs:14:20
+   |
+LL |     let b: usize;
+   |            ----- expected due to this type
+...
+LL |         res_field: b,
+   |                    ^ expected `usize`, found `Result<usize, ()>`
+   |
+   = note: expected type `usize`
+              found enum `Result<usize, ()>`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

